batchwrite.go

```diff
@@ -61,13 +62,174 @@ func (bw *BatchWrite) ConsumedCapacity(cc *ConsumedCapacity) *BatchWrite {
+	Error        error
+	BatchCounter int
+	Wrote        int
+}
+
+// Config used when calling RunConcurrently
+type batchWriteConfig struct {
+	threads int
+}
+
+// Parameter type to be passed to RunConcurrently
+type BatchWriteOption func(*batchWriteConfig)
+
+// Sets the default config
+func defaults(cfg *batchWriteConfig) {
+	cfg.threads = int(^uint(0) >> 1)
+}
+
+// Sets the number of threads to process the request
+func WithThreards(threads int) BatchWriteOption {
+	return func(cfg *batchWriteConfig) {
+		cfg.threads = threads
+	}
+}
+
+func (bw *BatchWrite) writeBatch(ctx aws.Context, ops []*dynamodb.WriteRequest, batchCounter int, channel chan<- BatchResponse) {
+
+	boff := backoff.WithContext(backoff.NewExponentialBackOff(), ctx)
+	wrote := 0
+
+	for {
+		var res *dynamodb.BatchWriteItemOutput
+		req := bw.input(ops)
+		err := retry(ctx, func() error {
+			var err error
+			res, err = bw.batch.table.db.client.BatchWriteItemWithContext(ctx, req)
+			return err
+		})
+		if err != nil {
+			channel <- BatchResponse{
+				Result:       res,
+				Error:        err,
+				BatchCounter: batchCounter,
+				Wrote:        0,
+			}
+			return
+		}
+		if bw.cc != nil {
+			for _, cc := range res.ConsumedCapacity {
+				addConsumedCapacity(bw.cc, cc)
+			}
+		}
+
+		unprocessed := res.UnprocessedItems[bw.batch.table.Name()]
+		wrote = len(ops) - len(unprocessed)
+		if len(unprocessed) == 0 {
+			channel <- BatchResponse{
+				Result:       res,
+				Error:        err,
+				BatchCounter: batchCounter,
+				Wrote:        wrote,
+			}
+			return
+		}
+		ops = unprocessed
+
+		// need to sleep when re-requesting, per spec
+		if err := aws.SleepWithContext(ctx, boff.NextBackOff()); err != nil {
+			channel <- BatchResponse{
+				Result:       nil,
+				Error:        err,
+				BatchCounter: batchCounter,
+				Wrote:        wrote,
+			}
+			return
+		}
+	}
+}
+
+func splitBatches(requests []*dynamodb.WriteRequest) (batches [][]*dynamodb.WriteRequest) {
+	batches = [][]*dynamodb.WriteRequest{}
+	requestsLength := len(requests)
+	for i := 0; i < requestsLength; i += maxWriteOps {
+		end := i + maxWriteOps
+		if end > requestsLength {
+			end = requestsLength
+		}
+		batches = append(batches, requests[i:end])
+	}
+	return batches
+}
+
+func min(a int, b int) int {
+
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// RunConcurrently executes this batch concurrently with the number of threads specified.
+// By default, it uses 2147483647 theads (if even possible) in an attempt to run each batch in parallel
+// In case of multiple errors, it will return only the last one.
+func (bw *BatchWrite) RunConcurrently(opts ...BatchWriteOption) (wrote int, err error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	return bw.RunConcurrentlyWithContext(ctx, opts...)
+}
+
+func (bw *BatchWrite) RunConcurrentlyWithContext(ctx aws.Context, opts ...BatchWriteOption) (wrote int, err error) {
+	if bw.err != nil {
+		return 0, bw.err
+	}
+
+	cfg := new(batchWriteConfig)
```

Comment:
You're not really using the notion of your threads really. I think it's overly complicating your operation as it's currently implemented. All you are really doing here is spinning up a go routine for each batch you're running, and you'll make too many operation if your threads number is larger than your total batches.

 If you were truly concerned about how many go routines were concurrently running, you'd be making a worker pool, like https://gobyexample.com/worker-pools and send your batches to be processed by the worker pool. I'd also not call these things "Threads" as a common Go convention would be to refer to it as "Pool Size".

Reply:
Will try this. Thanks!